streams_map.go

```diff
@@ -216,21 +217,50 @@ func (m *streamsMap) AcceptStream() (*stream, error) {
 	return str, nil
 }
 
-func (m *streamsMap) Iterate(fn streamLambda) error {
+func (m *streamsMap) DeleteClosedStreams() error {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
-	openStreams := append([]protocol.StreamID{}, m.openStreams...)
+	var numDeletedStreams int
+	// for every closed stream, the streamID is replaced by 0 in the openStreams slice
+	for i, streamID := range m.openStreams {
+		str, ok := m.streams[streamID]
+		if !ok {
+			return errMapAccess
+		}
+		if !str.finished() {
+			continue
+		}
+		m.removeStreamCallback(streamID)
+		numDeletedStreams++
+		m.openStreams[i] = 0
+		if streamID%2 == 0 {
+			m.numOutgoingStreams--
+		} else {
+			m.numIncomingStreams--
+		}
+		delete(m.streams, streamID)
+	}
 
-	for _, streamID := range openStreams {
-		cont, err := m.iterateFunc(streamID, fn)
-		if err != nil {
-			return err
+	if numDeletedStreams == 0 {
+		return nil
+	}
+
+	// remove all 0s (representing closed streams) from the openStreams slice
+	// and adjust the roundRobinIndex
+	var j int
+	for i, id := range m.openStreams {
+		if i != j {
+			m.openStreams[j] = m.openStreams[i]
 		}
-		if !cont {
-			break
+		if id != 0 {
+			j++
+		} else if uint32(j) < m.roundRobinIndex {
```

Comment:
Should we change the roundRobinIndex to an int?

Reply:
Yes.